### PR TITLE
Add device tree support to eth_mcux.c 

### LINKED ISSUE #10155: Add device tree support to eth_mcux.c
Look at adding device tree support for eth_mcux driver/device.  The #10097 has a reasonable example of how to do this.

### PR DESCRIPTION
**Pull request by agansari**
drivers: eth_mcux: replaces Kconfig for NXP MCUX Ethernet driver with Device Tree descriptor.
Fixes #10155 inspired from #10097 pull request.

### CODE REVIEW

**Review by galak — CHANGES_REQUESTED** [review]
Thanks for your first submission on this.  A few non-technical things:

Please take a few minutes to read over:

https://docs.zephyrproject.org/latest/contribute/contribute.html

It documents in greater detail some of the commit style guidelines we have for the project.  Some simple things are:

1. Having a Signed-off-by:
2. Having first line be a short description, with a subsystem identifier, something like:

`drivers: eth_mcux: Made some awesome changes`

3. Need a body describing in more detail the changes.

Please make these changes to the commits.

File: dts/arm/nxp/nxp_k6x.dtsi
Comment (galak) [inline dts/arm/nxp/nxp_k6x.dtsi]: What do we need port for?

File: dts/arm/nxp/nxp_k6x.dtsi
Comment (galak) [inline dts/arm/nxp/nxp_k6x.dtsi]: What do we need mtu for?

File: dts/arm/nxp/nxp_k6x.dtsi
Comment (galak) [inline dts/arm/nxp/nxp_k6x.dtsi]: We should probably have this be:
`local-mac-address = [00 00 00 00 00 00];`

and let the board dts file set it do a different value.

File: dts/bindings/ethernet/ethernet.yaml
Comment (galak) [inline dts/bindings/ethernet/ethernet.yaml]: Add license & copyright header info, see dts/bindings/gpio/gpio-leds.yaml for an example

File: dts/bindings/ethernet/nxp,mcux-ethernet.yaml
Comment (galak) [inline dts/bindings/ethernet/nxp,mcux-ethernet.yaml]: Add license & copyright header info, see dts/bindings/gpio/gpio-leds.yaml for an example

File: boards/arm/frdm_k64f/dts.fixup
Comment (galak) [inline boards/arm/frdm_k64f/dts.fixup]: Remove the #if ETH_PORT check, we tend to have the fixup always, if there isn't any reason not to.

**Review by galak — CHANGES_REQUESTED** [review]
Lets focus on the 'Initial DT implementation for eth_mcux' commit and get that clean and merged.  Lets hold off on the second commit 'MCUX Netorking configuration moved to Device Tree' for now.

Additional comments on the first commit.

File: boards/arm/frdm_k64f/dts.fixup
Comment (MaureenHelm) [inline boards/arm/frdm_k64f/dts.fixup]: It would be better to move the ethernet fixups to `soc/arm/nxp_kinetis/k6x/dts.fixup`

File: drivers/ethernet/Kconfig.mcux
Comment (MaureenHelm) [inline drivers/ethernet/Kconfig.mcux]: Need to keep this config, the others can be removed

File: dts/bindings/ethernet/nxp,mcux-ethernet.yaml
Comment (MaureenHelm) [inline dts/bindings/ethernet/nxp,mcux-ethernet.yaml]: This is describing the hardware, whereas mcux is a software concept. Change the constraint and filename to "nxp,kinetis-ethernet"

File: drivers/ethernet/Kconfig.mcux
Comment (MaureenHelm) [inline drivers/ethernet/Kconfig.mcux]: These are all driver configurations and should stay here, not move to dts.

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: boards/arm/frdm_k64f/frdm_k64f.dts
Comment (pfalcon) [inline boards/arm/frdm_k64f/frdm_k64f.dts]: As github shows, please add an end-line after this line.

File: drivers/ethernet/Kconfig.mcux
Comment (pfalcon) [reply drivers/ethernet/Kconfig.mcux]: +1

File: drivers/ethernet/Kconfig.mcux
Comment (pfalcon) [inline drivers/ethernet/Kconfig.mcux]: This select should stay (as many other settings too).

**Review by pfalcon — CHANGES_REQUESTED** [review]
@agansari: Thanks for updates so far.  Btw, it may be a good idea to to add a comment when you push a new version, because otherwise there's no notification on this (notifications go only on comments).

As was mentioned by @galak and @MaureenHelm already, this patch try to do too much at once, including moving things to DT which shouldn't be moved there (at least not right away). A good suggestion I may share in this regard: please try to do as little as possible for starters. It's much easier to add another thing if requested, then to throw away half of your work :-(. In this case (as was probably already mentioned), moving just base address and IRQ settings to DT is a solid good start. If done concisely, we can merge that, and start on another iteration to see if there's something else worth moving to DT (or not).

Thanks!

**Comment by MaureenHelm** [comment]
> In this case (as was probably already mentioned), moving just base address and IRQ settings to DT is a solid good start. If done concisely, we can merge that, and start on another iteration to see if there's something else worth moving to DT (or not).

In hindsight I see I wasn't very clear about which configs should move to dts and which should stay in kconfig. I agree with @pfalcon that base address and IRQs are a good start to move to dts. The IRQs are an important piece for this peripheral, because that's where the hardware instantiation differs on Kinetis K64 vs. i.MX RT1050 (the latter combines a couple of the IRQs). As a result, you may also want to add an `interrupt-names` attribute, since they will be different on the two SoCs.

If @galak wants to move local-mac-address to dts too, that's ok with me. But the others, like buffers and promiscuous mode, should stay in kconfig.

**Comment by agansari** [comment]
@galak @MaureenHelm @pfalcon I've made the changes you guys requested, can you have a second look?

File: drivers/ethernet/Kconfig.mcux
Comment (jukkar) [inline drivers/ethernet/Kconfig.mcux]: Why removing the mac address from Kconfig? If one wants to have a fixed MAC address, then one is forced to edit a file in version control which is not very convenient, or did I understood something wrong here.

Comment (agansari) [reply drivers/ethernet/Kconfig.mcux]: @jukkar The MAC is currently set in Kconfig to UNIQUE mode, i.e. implicit value. Manual MAC configured in DT is used only if Kconfig has set ETH_MCUX_0_MANUAL_MAC.

Comment (jukkar) [reply drivers/ethernet/Kconfig.mcux]: Sure, my point was that if one wants to set the MAC address, then one needs to edit the DT file which is under version control instead of putting the MAC address into a config file that is probably not under version control. Or did I misunderstood something here.

Comment (agansari) [reply drivers/ethernet/Kconfig.mcux]: Kconfig file is also under version control (file in discussion). I was pointing out that if you want to change the MAC manually then you need to make more changes and dt is not enough.

Comment (jukkar) [reply drivers/ethernet/Kconfig.mcux]: > Kconfig file is also under version control (file in discussion)

Sure, Kconfig file is but that does not contain the actual values of the MAC address as that is placed to the prj*conf. User is free to select his own conf file during compilation.

Not sure what do you mean by "...then you need to make more changes and dt is not enough."

Comment (agansari) [reply drivers/ethernet/Kconfig.mcux]: > Not sure what do you mean by "...then you need to make more changes and dt is not enough."

I was pointing out that with this change MAC should be changed via device tree.

**Comment by pfalcon** [comment]
@agansari :

> The MAC is currently set in Kconfig to UNIQUE mode

Thanks for ensuring this default is preserved with this refactor. (We went thru a bunch of confusion and pain to come to that default (https://github.com/zephyrproject-rtos/zephyr/issues/3187), so it really should stay like that).

File: drivers/ethernet/Kconfig.mcux
Comment (pfalcon) [reply drivers/ethernet/Kconfig.mcux]: @jukkar:

> Sure, my point was that if one wants to set the MAC address, then one needs to edit the DT file which is under version control

I find myself recently quoting @nashif on multiple times who said that we have [DT overlays](https://docs.zephyrproject.org/latest/application/application.html#application-dt). But I yet need to try myself how cumbersome it is. One point to have is that we won't get around need to override DT settings, so the earlier we start get used to it, the sooner we can may it more streamlined (if it's not such already).

I have mixed feelings about switching to `local-mac-address` in DT myself. But it's worth to point that it at least not our invention, it comes from Linux: https://elixir.bootlin.com/linux/latest/source/Documentation/devicetree/bindings/net/ethernet.txt . And if we need to go thru pain of it, then at least one can imagine a benefit: we have quite adhoc configs for manual MACs in various drivers, DT would standardize somewhat.

How much - depends, and we (those of us who don't know DT's ins and out) yet need to see. For example, will it be possible to have device-independent DT overlay which would specify MAC for any ethernet device? (E.g. using "eth" alias, as this patch introduces.)

All in all, I'm going to approve this, as at least it all makes sense and seems to work, while keeping the good defaults we have.

File: drivers/ethernet/Kconfig.mcux
Comment (pfalcon) [inline drivers/ethernet/Kconfig.mcux]: I'm not sure if I commented already or not, but removal of this line seems like a stray change, it rather stay.

**Review by pfalcon — APPROVED** [review]
A minor cosmetic fix (revert) suggestion, and otherwise looks good to me. (I looked at the generated configs/fixups to make sure previous UNIQUE default stays, while manual MAC from DT would actually be used by the code if such an options is selected. Didn't run on the board though.)

File: boards/arm/frdm_k64f/frdm_k64f.dts
Comment (galak) [inline boards/arm/frdm_k64f/frdm_k64f.dts]: What is this used for?  seems like we should remove it.

File: boards/arm/frdm_k64f/frdm_k64f.dts
Comment (galak) [inline boards/arm/frdm_k64f/frdm_k64f.dts]: Is there a reason for the ifdef CONFIG_PTP_CLOCK_MCUX?

File: boards/arm/frdm_k64f/frdm_k64f.dts
Comment (galak) [inline boards/arm/frdm_k64f/frdm_k64f.dts]: The board should enable ethernet by default, so please remove #ifdef CONFIG_NETWORKING

File: dts/bindings/ethernet/ethernet.yaml
Comment (galak) [inline dts/bindings/ethernet/ethernet.yaml]: remove id: (was recently removed from all binding files)

File: dts/bindings/ethernet/nxp.kinetis-ptp.yaml
Comment (galak) [inline dts/bindings/ethernet/nxp.kinetis-ptp.yaml]: remove id

File: soc/arm/nxp_kinetis/k6x/dts_fixup.h
Comment (galak) [inline soc/arm/nxp_kinetis/k6x/dts_fixup.h]: remove the ifdef

**Review by galak — CHANGES_REQUESTED** [review]
Some cleanup based on upstream changes and removing of #ifdef's

File: boards/arm/frdm_k64f/frdm_k64f.dts
Comment (pfalcon) [reply boards/arm/frdm_k64f/frdm_k64f.dts]: My understanding that this is kinda intended to select the default networking interface.

Yeah, I don't know how that works, if at all. It looks useful, but if not really implemented, could be postponed until later indeed.

**Comment by pfalcon** [comment]
@agansari, you also probably need to rebase this on master, as the PR for example contains renamed boards/x86/qemu_x86/dts.fixup → boards/x86/qemu_x86/dts_fixup.h. That's already done in master. Surprisingly, it doesn't show as conflict, but I'm sure it will lead to conflict on merge.

File: boards/arm/frdm_k64f/frdm_k64f.dts
Comment (pfalcon) [reply boards/arm/frdm_k64f/frdm_k64f.dts]: Maybe a better way to say it would be: regardless of whether networking is enabled or not, the eth hardware is there (routed to Eth socket, etc.). If CONFIG_NETWORKING is not enabled on Zephyr's side, it (hw) just won't be used. But there's no need to heavily #ifdef the DTS itself. Same applies to PTP.

(That's just my guess, as I'm learning DTS as we go with all these recent patches. So, if you could ack/nak it, @galak, it would be helpful.)

File: boards/arm/frdm_k64f/frdm_k64f.dts
Comment (galak) [reply boards/arm/frdm_k64f/frdm_k64f.dts]: Yeah, there's no support for this in any generation scripts, and I'm not sure what it would end up doing.  So we should remove for now.

File: boards/arm/frdm_k64f/frdm_k64f.dts
Comment (galak) [reply boards/arm/frdm_k64f/frdm_k64f.dts]: ack, if networking is disabled in Zephyr from a SW perspective, the hardware will not doing anything so there isn't a reason to protect it in the dts this way.

File: drivers/ethernet/Kconfig.mcux
Comment (jukkar) [reply drivers/ethernet/Kconfig.mcux]: Setting things via conf file feels more natural to me that doing things via some overlay DT. Anyway, I have no strong opinion on this and can for sure use some other board if frdm becomes too cumbersome to use in real life.

Comment (pfalcon) [reply drivers/ethernet/Kconfig.mcux]: > and can for sure use some other board if frdm becomes too cumbersome to use in real life.

Nah-nah, the whole point of this is that everything gets converted ;-). For example, recently @aescolar complained (bragged?) that native_posix doesn't use DT. I'm sure, it will be converted too.

**Comment by pfalcon** [comment]
@agansari: Thanks for updates. For a simple enough patches like this one, you can squash the changes into the main commit right away (that would need to be done before merging anyway).

**Comment by agansari** [comment]
Thank you @pfalcon @jukkar @galak for your reviews.
I've done another iteration of this pull request.

**Comment by pfalcon** [comment]
@galak, @MaureenHelm : Ping.

File: soc/arm/nxp_kinetis/k6x/dts_fixup.h
Comment (galak) [inline soc/arm/nxp_kinetis/k6x/dts_fixup.h]: Can you make these CONFIG_IRQ... (we need to update any references in the driver)

**Review by galak — CHANGES_REQUESTED** [review]
Looks good, one minor change to use CONFIG_IRQ_..

We've tried to keep with the convention for now to make it easy to know what is coming from DT to use CONFIG_.. for defines.

**Review by MaureenHelm — APPROVED** [review]

**Review by nashif — APPROVED** [review]

File: soc/arm/nxp_kinetis/k6x/dts_fixup.h
Comment (agansari) [reply soc/arm/nxp_kinetis/k6x/dts_fixup.h]: Hey @galak  I've made the change.

**Review by galak — APPROVED** [review]

**Pull request merged**
